### DISCUSSION TITLE
RAUC bundle metadata

### DISCUSSION THREAD

**Discussion by bettio** [rauc/rauc#888] [General]
We are working to implement a web application that allows to update a RAUC bundle and to store it on a bucket.
We are trying to make it less error prone as possible, so I have some questions:
* How we can reject a malformed or invalid RAUC bundle file? Is it possible to check some kind of magic or embedded manifest?
* How we can extract the compatible or any other metadata from the RAUC bundle file?
* Is it possible to embed any kind of additional metadata? such as an UUID, multilingual description or anything else?

**Comment by ejoerns** [comment]
> We are working to implement a web application that allows to update a RAUC bundle and to store it on a bucket. We are trying to make it less error prone as possible, so I have some questions:
> 

Not sure if I got the use case. Do you plan to modify a bundle in a web app?

> * How we can reject a malformed or invalid RAUC bundle file? Is it possible to check some kind of magic or embedded manifest?
> * How we can extract the compatible or any other metadata from the RAUC bundle file?

`rauc info --output-format=json bundle.raucb` should both reject an invalid bundle as well as dump basic manifest information in machine-readable format.

> * Is it possible to embed any kind of additional metadata? such as an UUID, multilingual description or anything else?

Not in the manifest. RAUC will reject bundles containing manifests with unknown metadata at the moment. What is the use case for adding this?

**Comment by jluebbe** [comment]
Hi,

> We are working to implement a web application that allows to update a RAUC bundle and to store it on a bucket. 

that sounds interesting! Is it / will it be open source?

In which cases would you want to modify a bundle after it was created and tested? The common workflow is that the bundle is generated reproducibly by a build-system (like Yocto), tested and then released. Any later modification would invalidate the reproducibility/testing and so should be avoided.

> We are trying to make it less error prone as possible, so I have some questions:

I'll respond in a different order.

> * How we can extract the compatible or any other metadata from the RAUC bundle file?

You can use `rauc info`, which supports json output:
```
$ ./rauc info --keyring=test/openssl-ca/dev-ca.pem --output-format=json test/good-verity-bundle.raucb | jq .
rauc-Message: 19:46:53.540: Reading bundle: /home/jluebbe/ptx/rauc/test/good-verity-bundle.raucb
rauc-Message: 19:46:53.541: Verifying bundle signature... 
rauc-Message: 19:46:53.542: Verified inline signature by 'O = Test Org, CN = Test Org Release-1'
{
  "compatible": "Test Config",
  "version": "2011.03-2",
  "description": null,
  "build": null,
  "hooks": [],
  "images": [
    {
      "rootfs": {
        "variant": null,
        "filename": "rootfs.img",
        "checksum": "101a4fc5c369a5c89a51a61bcbacedc9016e9510e59a4383f739ef55521f678d",
        "size": 8192,
        "hooks": []
      }
    },
    {
      "appfs": {
        "variant": null,
        "filename": "appfs.img",
        "checksum": "f95c0891937265df18ff962869b78e32148e7e97eab53fad7341536a24242450",
        "size": 8192,
        "hooks": []
      }
    }
  ]
}
```

If any relevant information is missing, it could be added there.

> * Is it possible to embed any kind of additional metadata? such as an UUID, multilingual description or anything else?

Currently, RAUC will reject unknown sections and keys in the manifest. One could add a specific section (or sections) which would be ignored during checks and only forwarded to `rauc info`. 

Could you give an example of the keys and values you'd want to add, and what they'd be useful for?

> * How we can reject a malformed or invalid RAUC bundle file?

As all RAUC bundles are signed, any unauthorized modification of the bundle file should be detected during installation. `rauc extract` will also verify the bundle payload before starting the extraction. Note that you need to have the CA certificate (`--keyring`) available to check the signature. This also means that if you want to modify a bundle, you basically have to extract, modify and create a new bundle. 

I'm not sure how a bundle could have a valid signature but be "malformed". Bundles should only be created by `rauc bundle`, which will check and update the manifest as needed and then sign it. The contents of the bundle can't really be checked automatically, especially if you don't know the target's configuration (`system.conf` and related configuration).

> Is it possible to check some kind of magic or embedded manifest?

You can read the CMS signature at the end of the bundle, which (for verity and crypt bundles) contains the manifest. By checking the signature, you can be sure that it was produced by a specific trusted entity.

**Reply by bettio** [reply]
> > We are working to implement a web application that allows to update a RAUC bundle and to store it on a bucket.
> 
> that sounds interesting! Is it / will it be open source?

Yes, that functionality is part of the wider [Edgehog Device Manager](https://github.com/edgehog-device-manager) project.
 
> In which cases would you want to modify a bundle after it was created and tested? The common workflow is that the bundle is generated reproducibly by a build-system (like Yocto), tested and then released. Any later modification would invalidate the reproducibility/testing and so should be avoided.

Sorry, I mistyped, I actually meant "a web application that allows to *upload a RAUC bundle and to store it on a bucket."

> > * How we can extract the compatible or any other metadata from the RAUC bundle file?
> 
> You can use `rauc info`, which supports json output:

Looks great, however that will require running the rauc tool on the backend service, do you think that would be easily feasible extracting those information using some kind of library or by parsing any section of the raucb file? If I understood correctly you mentioned the CMS section, so I just need to find its offset, right?

> > * Is it possible to embed any kind of additional metadata? such as an UUID, multilingual description or anything else?
> 
> Currently, RAUC will reject unknown sections and keys in the manifest. One could add a specific section (or sections) which would be ignored during checks and only forwarded to `rauc info`.

Sounds interesting, can you give me a short example that would work without any compatibility issue (also with future releases, I don't want to go "out of spec")?

> Could you give an example of the keys and values you'd want to add, and what they'd be useful for?

We've been thinking about information such as multiple localized bundle descriptions and identification UUID. 

> > * How we can reject a malformed or invalid RAUC bundle file?
> 
> As all RAUC bundles are signed, any unauthorized modification of the bundle file should be detected during installation. `rauc extract` will also verify the bundle payload before starting the extraction. Note that you need to have the CA certificate (`--keyring`) available to check the signature. This also means that if you want to modify a bundle, you basically have to extract, modify and create a new bundle.

Yes, that's clear. Sorry for having mistyped.

> I'm not sure how a bundle could have a valid signature but be "malformed". Bundles should only be created by `rauc bundle`, which will check and update the manifest as needed and then sign it. The contents of the bundle can't really be checked automatically, especially if you don't know the target's configuration (`system.conf` and related configuration).

We would like to:
- avoid storing online invalid or malformed bundles or to announce them to remote devices.
- ensuring that the target group of devices for which the bundle is uploaded is compatible with it, we acknowledge that devices are able to reject the bundle once is downloaded, however we would rather return an error as soon as it is uploaded to the web application.  

Hence it would be great to extract metadata as soon as the bundle is uploaded. 

> > Is it possible to check some kind of magic or embedded manifest?
> 
> You can read the CMS signature at the end of the bundle, which (for verity and crypt bundles) contains the manifest. By checking the signature, you can be sure that it was produced by a specific trusted entity.

The whole manifest or just a subset of it? Also, is the CMS signature extern to the squashfs filesystem?
